Add support for Shrink ONNX operator

gyulaz-htec opened the pull request:
Note: depends on #2179, until it's merged marked as Draft

This PR adds support for Shrink ONNX operator.

Resolves https://github.com/migraphx-benchmark/AMDMIGraphX/issues/118

pfultz2 commented:
A new migraphx operator should not be created for this. This should use existing operators that should be inserted during parse_onnx.

gyulaz-htec commented:
Thanks @pfultz2 I've changed the PR according to that.

umangyadav commented on src/onnx/parse_shrink.cpp:
```diff
@@ -0,0 +1,102 @@
+        instruction_ref mb_neg_lambd;
+        instruction_ref mb_lambd;
+        instruction_ref mb_zero;
+        if(x_shape.dynamic())
```
You can use `add_common_op` instead. 
That would handle broadcasting and conversion of data type for you. 
https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/e5cd8b6b0ec30f7a3b7b6775b096c3a41003b09d/src/onnx/onnx_parser.cpp#L172

umangyadav replied:
bias and lambda is only of dtype `float` but input can have dtype other than float. Therefore need to handle conversions as well. Can you add tests for those as well ?

gyulaz-htec replied:
I've added onnx test cases for these.

CharlieL7 commented on src/onnx/parse_shrink.cpp:
```diff
@@ -0,0 +1,102 @@
+        auto x_plus_bias = info.add_instruction(migraphx::make_op("add"), args[0], mb_bias);
+        auto x_min_bias  = info.add_instruction(migraphx::make_op("sub"), args[0], mb_bias);
+
+        auto filtered =
+            info.add_instruction(migraphx::make_op("where"), condition_1, x_plus_bias, mb_zero);
+        return info.add_instruction(migraphx::make_op("where"), condition_2, x_min_bias, filtered);
```
ONNX spec is unclear to me if they want:
```
if(x < -lambd){ y = x + bias )
if(x > lambd){ y = x -bias }
else { y = 0}
```
or
```
if(x < -lambd){ y = x + bias )
else if(x > lambd){ y = x -bias }
else { y = 0}
```
It looks like this implements the `else if` version

gyulaz-htec replied:
Good question for example `x=0` and `lamd=-5` can pass both conditions . I will check the reference implementation and come back with an answer.

gyulaz-htec replied:
onnxruntime uses the `if`,`if`,`else` solution: https://github.com/microsoft/onnxruntime/blob/25bbd8d4ebed60c28f702a9fe6b0e6a12c19de9c/onnxruntime/core/providers/cpu/nn/shrink.cc#L37-L47

gyulaz-htec commented:
@umangyadav I've added the requested test cases to check the conversion handling.
@CharlieL7 as I mentioned [above](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/pull/2240#discussion_r1356896600) the onnxruntime uses the `if elseif else` approoach, but my previous solution had some flaws and didn't handle overlapping conditions properly.
I've fixed that and also added two verify test cases where the first two conditions can overlap, but the result is picked from the first condition.

CharlieL7 commented on src/onnx/parse_shrink.cpp:
```diff
@@ -0,0 +1,85 @@
+
+        auto first  = info.add_common_op("mul", mul1, x_plus_bias);
+        auto second = info.add_common_op("mul", mul2, x_min_bias);
+        auto ret    = info.add_common_op("add", first, second);
```
It looks like it could, but there's no need for a tensor of zeroes this way?

umangyadav replied:
`where` can only do binary classification. 
Here it has `if`, `else if` and `else`. that's why where can't be used alone i think.

umangyadav replied:
> It looks like it could, but there's no need for a tensor of zeroes this way? 

yeah.

causten commented:
There is a CI failure

causten commented:
[   RUN    ] test_shrink<migraphx::shape::uint64_type>

/usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_algobase.h:342:20: runtime error: -3 is outside the range of representable values of type 'unsigned long'

    #0 0xf3c8a0 in migraphx::version_1::basic_iota_iterator<migraphx::version_1::tensor_view_iterator_read<migraphx::version_1::tensor_view<unsigned long>>, unsigned long> std::__copy_move<false, false, std::random_access_iterator_tag>::__copy_m<float const*, migraphx::version_1::basic_iota_iterator<migraphx::version_1::tensor_view_iterator_read<migraphx::version_1::tensor_view<unsigned long>>, unsigned long>>(float const*, float const*, migraphx::version_1::basic_iota_iterator<migraphx::version_1::tensor_view_iterator_read<migraphx::version_1::tensor_view<unsigned long>>, unsigned long>) /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_algobase.h:342:20

    #1 0xf39174 in migraphx::version_1::basic_iota_iterator<migraphx::version_1::tensor_view_iterator_read<migraphx::version_1::tensor_view<unsigned long>>, unsigned long> std::__copy_move_a<false, float const*, migraphx::version_1::basic_iota_iterator<migraphx::version_1::tensor_view_iterator_read<migraphx::version_1::tensor_view<unsigned long>>, unsigned long>>(float const*, float const*, migraphx::version_1::basic_iota_iterator<migraphx::version_1::tensor_view_iterator_read<migraphx::version_1::tensor_view<unsigned long>>, unsigned long>) /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_algobase.h:403:14

    #2 0xf39174 in migraphx::version_1::basic_iota_iterator<migraphx::version_1::tensor_view_iterator_read<migraphx::version_1::tensor_view<unsigned long>>, unsigned long> std::__copy_move_a2<false, __gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>, migraphx::version_1::basic_iota_iterator<migraphx::version_1::tensor_view_iterator_read<migraphx::version_1::tensor_view<unsigned long>>, unsigned long>>(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>, __gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>, migraphx::version_1::basic_iota_iterator<migraphx::version_1::tensor_view_iterator_read<migraphx::version_1::tensor_view<unsigned long>>, unsigned long>) /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_algobase.h:441:3

    #3 0xf39174 in migraphx::version_1::basic_iota_iterator<migraphx::version_1::tensor_view_iterator_read<migraphx::version_1::tensor_view<unsigned long>>, unsigned long> std::copy<__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>, migraphx::version_1::basic_iota_iterator<migraphx::version_1::tensor_view_iterator_read<migraphx::version_1::tensor_view<unsigned long>>, unsigned long>>(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>, __gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>, migraphx::version_1::basic_iota_iterator<migraphx::version_1::tensor_view_iterator_read<migraphx::version_1::tensor_view<unsigned long>>, unsigned long>) /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_algobase.h:473:14

    #4 0xf39174 in auto void migraphx::version_1::literal::fill<__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>>(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>, __gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>)::'lambda'(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>)::operator()<migraphx::version_1::shape::as<unsigned long>>(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>) const /var/jenkins/workspace/AMDMIGraphX_PR-2240/src/include/migraphx/literal.hpp:118:13

    #5 0xf35c15 in void migraphx::version_1::shape::visit<void migraphx::version_1::literal::fill<__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>>(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>, __gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>)::'lambda'(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>), void migraphx::version_1::shape::visit<void migraphx::version_1::literal::fill<__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>>(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>, __gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>)::'lambda'(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>)>(migraphx::version_1::shape::type_t, __gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>)::'lambda'()>(migraphx::version_1::shape::type_t, __gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>, void migraphx::version_1::shape::visit<void migraphx::version_1::literal::fill<__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>>(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>, __gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>)::'lambda'(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>)>(migraphx::version_1::shape::type_t, __gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>)::'lambda'()) /var/jenkins/workspace/AMDMIGraphX_PR-2240/src/include/migraphx/shape.hpp:366:13

    #6 0xf359d9 in void migraphx::version_1::shape::visit<void migraphx::version_1::literal::fill<__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>>(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>, __gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>)::'lambda'(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>)>(migraphx::version_1::shape::type_t, __gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>) /var/jenkins/workspace/AMDMIGraphX_PR-2240/src/include/migraphx/shape.hpp:375:16

    #7 0xf359d9 in void migraphx::version_1::shape::visit_type<void migraphx::version_1::literal::fill<__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>>(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>, __gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>)::'lambda'(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>)>(void migraphx::version_1::literal::fill<__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>>(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>, __gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>)::'lambda'(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>)) const /var/jenkins/workspace/AMDMIGraphX_PR-2240/src/include/migraphx/shape.hpp:381:9

    #8 0xf359d9 in void migraphx::version_1::literal::fill<__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>>(__gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>, __gnu_cxx::__normal_iterator<float const*, std::vector<float, std::allocator<float>>>) /var/jenkins/workspace/AMDMIGraphX_PR-2240/src/include/migraphx/literal.hpp:116:17

    #9 0xf35799 in migraphx::version_1::literal::literal<float>(migraphx::version_1::shape const&, std::vector<float, std::allocator<float>> const&) /var/jenkins/workspace/AMDMIGraphX_PR-2240/src/include/migraphx/literal.hpp:65:9

    #10 0x19c8b9f in test_shrink<(migraphx::version_1::shape::type_t)11>::create_program() const /var/jenkins/workspace/AMDMIGraphX_PR-2240/test/verify/test_shrink.cpp:44:41

    #11 0x19c8695 in void register_verify_program_action::apply<test_shrink<(migraphx::version_1::shape::type_t)11>>()::'lambda'()::operator()() const /var/jenkins/workspace/AMDMIGraphX_PR-2240/test/verify/verify_program.hpp:65:45

    #12 0x19c8695 in std::_Function_handler<migraphx::version_1::program (), void register_verify_program_action::apply<test_shrink<(migraphx::version_1::shape::type_t)11>>()::'lambda'()>::_M_invoke(std::_Any_data const&) /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/std_function.h:285:9

    #13 0x1243a2e in std::function<migraphx::version_1::program ()>::operator()() const /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/std_function.h:688:14

    #14 0x1243a2e in run_verify::run(int, char const**) const::$_0::operator()() const /var/jenkins/workspace/AMDMIGraphX_PR-2240/test/verify/run_verify.cpp:276:58

    #15 0x1243a2e in std::_Function_handler<void (), run_verify::run(int, char const**) const::$_0>::_M_invoke(std::_Any_data const&) /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/std_function.h:300:2

    #16 0x125cfb6 in std::function<void ()>::operator()() const /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/std_function.h:688:14

    #17 0x125cfb6 in test::driver::run_test_case(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>> const&, std::function<void ()> const&, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>>>, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>>, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>> const, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>>>>>> const&) /var/jenkins/workspace/AMDMIGraphX_PR-2240/test/verify/../include/test.hpp:601:17

    #18 0x125535e in test::driver::run(int, char const**) /var/jenkins/workspace/AMDMIGraphX_PR-2240/test/verify/../include/test.hpp:668:25

    #19 0x124333e in run_verify::run(int, char const**) const /var/jenkins/workspace/AMDMIGraphX_PR-2240/test/verify/run_verify.cpp:284:7

    #20 0xf7fb33 in main /var/jenkins/workspace/AMDMIGraphX_PR-2240/test/verify/main.cpp:81:8

    #21 0x7f1af6ca8082 in __libc_start_main /build/glibc-BHL3KM/glibc-2.31/csu/../csu/libc-start.c:308:16

    #22 0xde1a1d in _start (/var/jenkins/workspace/AMDMIGraphX_PR-2240/build/bin/test_verify+0xde1a1d)



SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_algobase.h:342:20 in 

CMake Error at gdb/test_test_verify_general/run.cmake:16 (message):

  Test failed

gyulaz-htec commented:
@causten the CI fail should be fixed now
